### DISCUSSION TITLE
Proper way of using type union with declared types

### DISCUSSION THREAD

**Discussion by ymc9** [eclipse-langium/langium#693] [Q&A] [answered]
Hi guys,

I'm new to Langium and excited trying it out. I'm trying to fully use declared types but not sure how to use type union properly. Here's a simple repro:

```
grammar Test

interface A {
    value: string;
}

interface B {
    value: number;
}

type AB = A | B;

entry Model:
    R;

R returns AB:
    {A} value=STRING | {B} value=INT;

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;
terminal STRING: /"[^"]*"|'[^']*'/;
terminal INT returns number: /[0-9]+/;
hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

Langium reports error for the declaration of "type AB" and rule "R":
Inferred and declared versions of type AB both have to be interfaces or unions.

I guess I must have missed something very obvious. Any help is appreciated. Thanks!

**Comment by msujew** [comment, marked as answer]
Hey @ymc9,

right now the whole declared types stuff is in a relatively early feature stage. There are [tons of issues](https://github.com/langium/langium/issues?q=is%3Aissue+is%3Aopen+label%3Atypes) in both the validation and generation of those types. We plan to finalize this feature in Q4, but as of now, I wouldn't recommend using it. Instead, you should just let Langium infer your types, that works fine in almost all cases.

**Reply by ymc9** [reply]
Thanks! I'll try out inferred type then.

Maybe it's a good idea to mention it in the doc here: https://langium.org/docs/ast-types/#declared-types. It reads like declared types is recommended.

**Reply by montymxb** [reply]
That's a good point. We can maybe add a small disclaimer there to keep new and interested users apprised.
